ts/components/BarcodeCamera.tsx

```diff
@@ -0,0 +1,176 @@
+import { Camera, useCameraDevices } from "react-native-vision-camera";
+import { View, Dimensions, StyleSheet, Vibration } from "react-native";
+import { useScanBarcodes, BarcodeFormat } from "vision-camera-code-scanner";
+import { Text } from "native-base";
+import I18n from "../i18n";
+import customVariables, {
+  VIBRATION_BARCODE_SCANNED_DURATION
+} from "../theme/variables";
+import { openAppSettings } from "../utils/appSettings";
+import ButtonDefaultOpacity from "./ButtonDefaultOpacity";
+
+/**
+ * Type describing the supported barcodes in IO.
+ */
+export type IOBarcodeFormat = "QRCODE";
+
+/**
+ * The message sent through the `onBarcodeScanned`
+ * callback.
+ */
+export type ScannedBarcode = {
+  format: IOBarcodeFormat;
+  value: string;
+};
+
+/**
+ * Map the barcode scanned through the underneath
+ * library to one of the supported ones in IO, or `null`.
+ */
+function barcodeFormatToIOFormat(
+  format: BarcodeFormat
+): IOBarcodeFormat | null {
+  switch (format) {
+    case BarcodeFormat.QR_CODE:
+      return "QRCODE";
+
+    default:
+      return null;
+  }
+}
+
+const screenWidth = Dimensions.get("screen").width;
+
+const styles = StyleSheet.create({
+  cameraContainer: {
+    position: "relative",
+    width: "100%",
+    height: screenWidth,
+    backgroundColor: "#000"
+  },
+
+  camera: {
+    position: "absolute",
+    width: "100%",
+    height: "100%"
+  },
+
+  notAuthorizedContainer: {
+    padding: customVariables.contentPadding,
+    flex: 1,
+    alignItems: "center",
+    alignSelf: "stretch",
+    marginBottom: 14
+  },
+
+  notAuthorizedText: {
+    marginBottom: 25
+  },
+
+  notAuthorizedBtn: {
+    flex: 1,
+    alignSelf: "stretch"
+  }
+});
+
+type Props = {
+  onBarcodeScanned: (barcode: ScannedBarcode) => void;
+  disabled?: boolean;
+  marker?: React.ReactNode;
+};
+
+/**
+ * A Camera view which can scan different types of
+ * barcodes (QRCodes, Data Matrix, ...).
+ */
+export const BarcodeCamera = (props: Props) => {
+  const { onBarcodeScanned, disabled } = props;
+  const devices = useCameraDevices();
+  const [permissionsGranted, setPermissionsGranted] = useState(false);
+  const device = devices.back;
+
+  const [frameProcessor, barcodes] = useScanBarcodes([BarcodeFormat.QR_CODE], {
+    checkInverted: true
+  });
+
+  // Hook that handles the permissions initialization.
+  useEffect(() => {
+    async function checkPermissions() {
+      const cameraPermissions = await Camera.getCameraPermissionStatus();
+
+      if (cameraPermissions === "not-determined") {
+        const selectedPermissions = await Camera.requestCameraPermission();
+        setPermissionsGranted(selectedPermissions === "authorized");
+      } else {
+        setPermissionsGranted(cameraPermissions === "authorized");
+      }
+    }
+
+    void checkPermissions();
+  }, [setPermissionsGranted]);
+
+  // Hook that handles the `onBarcodeScanned` callback.
+  useEffect(() => {
+    // This is going to take only the first scanned
```

Comment:
```suggestion
   if (disabled) {
      return;
    }
    // This is going to take only the first scanned
```
⚠️ I noticed it in Android: it continues to scan and raises the event `onBarcodeScanned` multiple times. In fact it vibrates for 50ms for around 4-5 times

Reply:
Yes, I replicated it too, this is the fix for this bug: https://github.com/pagopa/io-app/pull/3928/commits/f83df0cf2c64168b96f555eff28d5ed064a429fd.

I also noticed that in older devices there is a small race condition in which I scan a barcode, go back to the camera screen, and it automatically re-scan the same barcode as soon as I land in the previous screen. Just in older devices because it's probably related to the time used to initialise the scanner and the various hook dependencies in `useEffect`. This has been fixed creating an hook `usePrevious` which is quite standard in its implementation, it just save the "previous" value of a variable inside a function component, and using it to check the previous `disabled` value in order to trigger the scan only when the camera it's actually ready. This is the fix: https://github.com/pagopa/io-app/pull/3928/commits/d3bff7bfd6800fe5ecd83fb1655e8261c5e05efb.

Reply:
Great point!
`usePrevious` is a useful hook, spread its existence to the others: 🗣️